File: 7 Reverse Integer/7 LL.cpp
#include <iostream>
using namespace std;

/*
https://leetcode.com/problems/reverse-integer/

Reverse digits of an integer.

Example1: x = 123, return 321
Example2: x = -123, return -321
*/

class Solution
{
public:
	int reverse(int x)
	{
		if (x == ~0x7FFFFFFF)
		{
			return 0;
		}
		bool neg = (x < 0);
		x = neg ? -x : x;
		int digits[16];
		int pos = 0;
		while (x != 0)
		{
			digits[pos++] = x % 10;
			x /= 10;
		}
		long long int result = 0;
		for (int i = 0; i < pos; ++i)
		{
			result = result * 10 + digits[i];
		}
		result = (result > 0x7FFFFFFF) ? 0 : result;
		return neg ? -int(result) : int(result);
	}
};

/*
idea: careful handle overflow
complexity: O(1)
*/

int main(void)
{
	Solution engine;
	cout << engine.reverse(-123) << '\n';
	return 0;
}

File: 7 Reverse Integer/7 LL.py
class Solution(object):
	def reverse(self, x):
		if x == ~0x7FFFFFFF:
			return 0
		neg = (x < 0)
		x = -x if neg else x
		digits = list()
		while x > 0:
			digits.append(x % 10)
			x /= 10
		result = 0
		for i in range(0, len(digits)):
			result = result * 10 + digits[i]
		result = 0 if result > 0x7FFFFFFF else result
		return -result if neg else result

engine = Solution()
print engine.reverse(-123)